src/flint/types/fq_default.pyx

```diff
@@ -0,0 +1,886 @@
+            if fq_type is None:
+                raise ValueError("invalid fq_type string")
+
+        # Now fq_type should be an int between 0, 5
+        if not typecheck(fq_type, int):
+            raise TypeError(f"{fq_type = } is invalid")
+        if fq_type < 0 or fq_type > 5:
+            raise ValueError(f"{fq_type = } should be between 0 and 5")
+
+        return fq_type
+
+    @staticmethod
+    def _parse_input_var(var):
+        # If no variable is given, use x
+        if var is None:
+            var = b"x"
+
+        # Encode to bytes for cython to parse
+        if isinstance(var, str):
+            var = var.encode()
+
+        # TODO: Flint only wants one-character inputs
+        if len(var) > 1:
+            raise ValueError("variable for GF(p^k) generator can only be one character")
+
+        return var
+
+    def __init__(self, p=None, degree=None, var=None, modulus=None, fq_type=fq_default_type.DEFAULT):
+        # Ensure the var used for the generator of GF(p^d) is a single byte
+        # TODO: this var is meaningless for GF(p) -- we could handle this somehow?
+        var = self._parse_input_var(var)
+
+        # Ensure the fq_type is an integer between 0, 5 -- we allow users to
+        # input a string which is converted as an enum
+        fq_type = self._parse_input_fq_type(fq_type)
+
+        # If a modulus is given, attempt to construct from this
+        if modulus is not None:
+            # If the polynomial has no known characteristic, we can try and create one
+            # using the supplied prime
+            if not (typecheck(modulus, fmpz_mod_poly) or typecheck(modulus, nmod_poly)):
+                if p is None:
+                    raise ValueError("cannot create from modulus if no characteristic is known")
+
+                ring_ctx = fmpz_mod_poly_ctx(p)
+                modulus = ring_ctx.any_as_fmpz_mod_poly(modulus)
+                if modulus is NotImplemented:
+                    raise TypeError("modulus cannot be cast to fmpz_mod_poly")
+
+            self.set_from_modulus(modulus, var, fq_type)
+            return
+
+        # If there's no modulus and no prime, we can't continue
+        if p is None:
+            raise ValueError("either a prime or modulus must be passed for construction")
+
+        # If we're not given a degree, construct GF(p)
+        if degree is None:
+            degree = 1
+
+        # Construct the field from the prime and degree GF(p^d)
+        self.set_from_order(p, degree, var, fq_type)
+
+    cdef _c_set_from_order(self, fmpz p, int d, char *var,
+                           fq_default_type fq_type=fq_default_type.DEFAULT):
+        self.var = var
+        fq_default_ctx_init_type(self.val, p.val, d, self.var, fq_type)
+        self._initialized = True
+
+    def set_from_order(self, p, d, var,
+                       fq_type=fq_default_type.DEFAULT, check_prime=True):
+        """
+        Construct a context for the finite field GF(p^d).
```

Comment:
I don't think we want these context-mutating methods to be public. I assume that it would wreak havoc if someone mutated the context after creating elements...

Reply:
Maybe make them `cdef` only with leading underscore names.